syncer/syncer.go

```
@@ -0,0 +1,419 @@
+		s.setSyncState(ctx, notSynced)
+	}
+}
+
+func (s *Syncer) setStateAfterSync(ctx context.Context, success bool) {
+	if !success {
+		s.setSyncState(ctx, notSynced)
+		return
+	}
+	currSyncState := s.getSyncState()
+	current := s.ticker.GetCurrentLayer()
+	// if we have gossip-synced to the target synced layer, we are ready to participate in the consensus
+	if currSyncState == gossipSync && s.getTargetSyncedLayer() <= current {
+		s.setSyncState(ctx, synced)
+	} else if currSyncState == notSynced {
+		// wait till s.ticker.GetCurrentLayer() + numGossipSyncLayers to participate in consensus
+		s.setSyncState(ctx, gossipSync)
+		s.setTargetSyncedLayer(ctx, current+numGossipSyncLayers)
+	}
+}
+
+func (s *Syncer) syncLayer(ctx context.Context, layerID types.LayerID) (*types.Layer, error) {
+	if s.isClosed() {
+		return nil, errors.New("shutdown")
+	}
+
+	layer, err := s.getLayerFromPeers(ctx, layerID)
+	if err != nil {
+		s.logger.WithContext(ctx).With().Error("failed to get layer from neighbors", layerID, log.Err(err))
+		return nil, err
+	}
+
+	if err := s.getATXs(ctx, layerID); err != nil {
+		return nil, err
+	}
+	return layer, nil
+}
+
+func (s *Syncer) getLayerFromPeers(ctx context.Context, layerID types.LayerID) (*types.Layer, error) {
+	ch := s.fetcher.PollLayer(ctx, layerID)
+	res := <-ch
+	if res.Err != nil {
+		if res.Err == layerfetcher.ErrZeroLayer {
+			return types.NewLayer(layerID), nil
+		}
+		return nil, res.Err
+	}
+	return s.mesh.GetLayer(layerID)
+}
+
+func (s *Syncer) getATXs(ctx context.Context, layerID types.LayerID) error {
+	if layerID.GetEpoch() == 0 {
+		s.logger.WithContext(ctx).Info("skip getting ATX in epoch 0")
+		return nil
+	}
+	epoch := layerID.GetEpoch()
+	s.logger.WithContext(ctx).With().Info("getting ATXs", epoch, layerID)
+	ctx = log.WithNewRequestID(ctx, layerID.GetEpoch())
+	if err := s.fetcher.GetEpochATXs(ctx, layerID.GetEpoch()); err != nil {
+		// TODO: is this really expected? we should have ATXs in epoch 1
```

@antonlerner 

we already skip ATX for epch 0. should we ignore error for epoch 1? i'd imagine we will have ATX in epoch 1 when miners do the PoET work during epoch 0.

---

yes we need to ignore only in epoch 0

---

fixed. thanks